File: Tests/MicroKeysCoreTests/FrameDecoderTests.swift
import Testing
@testable import MicroKeysCore

private func report(_ json: String, prefixed: Bool = false, size: Int = 63) -> [UInt8] {
    let body = Array((json + "\r\n").utf8)
    var out: [UInt8] = prefixed ? [6] : []
    out += [0x02, UInt8(body.count)] + body
    while out.count < size { out.append(0) }
    return out
}

@Suite struct FrameDecoderTests {
    @Test func keyDownUSBFraming() {
        var d = FrameDecoder()
        let events = d.feed(report(#"{"m":"v.oai.hid","p":{"k":"ACT10","act":1,"ag":0}}"#))
        #expect(events == [.key(id: "ACT10", act: 1)])
    }

    @Test func keyUpBLEPrefixedFraming() {
        var d = FrameDecoder()
        let events = d.feed(report(#"{"m":"v.oai.hid","p":{"k":"AG03","act":0}}"#, prefixed: true, size: 64))
        #expect(events == [.key(id: "AG03", act: 0)])
    }

    @Test func messageSpanningTwoReports() {
        var d = FrameDecoder()
        let json = #"{"jsonrpc":"2.0","method":"v.oai.hid","params":{"k":"ENC_CW","act":2,"padding":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"}}"#
        let body = Array((json + "\r\n").utf8)
        let first = Array(body[..<60]), second = Array(body[60...])
        #expect(d.feed([0x02, UInt8(first.count)] + first) == [])
        #expect(d.feed([0x02, UInt8(second.count)] + second) == [.key(id: "ENC_CW", act: 2)])
    }

    @Test func twoMessagesInOneReport() {
        var d = FrameDecoder()
        let json = #"{"m":"v.oai.hid","p":{"k":"ACT06","act":1}}"# + "\r\n" + #"{"m":"v.oai.rad","p":{"a":0.5,"d":0.25}}"#
        let body = Array((json + "\r\n").utf8)
        #expect(d.feed([0x02, UInt8(body.count)] + body) ==
                [.key(id: "ACT06", act: 1), .joystick(angle: 0.5, distance: 0.25)])
    }

    @Test func garbageIsIgnored() {
        var d = FrameDecoder()
        #expect(d.feed([0x01, 0x00, 0x00]) == [])
        #expect(d.feed([]) == [])
        #expect(d.feed(report("not json")) == [])
    }
}

File: windows/MicroKeys.Core.Tests/FrameDecoderTests.cs
using MicroKeys.Core;
using Xunit;

namespace MicroKeys.Core.Tests;

public class FrameDecoderTests
{
    [Fact]
    public void KeyDownUsbFraming()
    {
        var d = new FrameDecoder();
        var events = d.Feed(Reports.Make("""{"m":"v.oai.hid","p":{"k":"ACT10","act":1,"ag":0}}"""));
        Assert.Equal(new PadEvent[] { new PadEvent.Key("ACT10", 1) }, events);
    }

    [Fact]
    public void KeyUpWithReportIdPrefix()
    {
        var d = new FrameDecoder();
        var events = d.Feed(Reports.Make("""{"m":"v.oai.hid","p":{"k":"AG03","act":0}}""", prefixed: true, size: 64));
        Assert.Equal(new PadEvent[] { new PadEvent.Key("AG03", 0) }, events);
    }

    [Fact]
    public void MessageSpanningTwoReports()
    {
        var d = new FrameDecoder();
        var json = """{"jsonrpc":"2.0","method":"v.oai.hid","params":{"k":"ENC_CW","act":2,"padding":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"}}""";
        var body = System.Text.Encoding.UTF8.GetBytes(json + "\r\n");
        var first = body[..60]; var second = body[60..];
        Assert.Empty(d.Feed(new byte[] { 0x02, (byte)first.Length }.Concat(first).ToArray()));
        Assert.Equal(new PadEvent[] { new PadEvent.Key("ENC_CW", 2) },
            d.Feed(new byte[] { 0x02, (byte)second.Length }.Concat(second).ToArray()));
    }

    [Fact]
    public void TwoMessagesInOneReport()
    {
        var d = new FrameDecoder();
        var json = """{"m":"v.oai.hid","p":{"k":"ACT06","act":1}}""" + "\r\n" + """{"m":"v.oai.rad","p":{"a":0.5,"d":0.25}}""";
        var body = System.Text.Encoding.UTF8.GetBytes(json + "\r\n");
        Assert.Equal(new PadEvent[] { new PadEvent.Key("ACT06", 1), new PadEvent.Joystick(0.5, 0.25) },
            d.Feed(new byte[] { 0x02, (byte)body.Length }.Concat(body).ToArray()));
    }

    [Fact]
    public void GarbageIsIgnored()
    {
        var d = new FrameDecoder();
        Assert.Empty(d.Feed(new byte[] { 0x01, 0x00, 0x00 }));
        Assert.Empty(d.Feed(Array.Empty<byte>()));
        Assert.Empty(d.Feed(Reports.Make("not json")));
    }
}
